add MonadTrans overview and MaybeT example

levinotik opened the pull request:
I've been given commit access, but would like a review of this new material first. Thank you

levinotik committed:
add MonadTrans overview and MaybeT example

levinotik commented:
Is the `lift` example contrived given that the example uses `IO`?

snoyberg commented on content/monad-transformers.md:
```diff
@@ -23,13 +23,57 @@ A `Reader r a` in which the resulting `a` is wrapped in any other monad, i.e. `r
+
+### MonadTrans
+
+[transformers](https://www.stackage.org/lts-3.20/package/transformers-0.4.2.0) is a widely used package which provides transformer versions of various monads. 
```
To show the most up-to-date version of the docs, I'd recommend making the link point to:

https://www.stackage.org/package/transformers

snoyberg commented:
I think the example is good. A note afterwards that, in the case of a one-layer transformer on top of `IO`, `lift = liftIO`, may address the concern you're raising.

levinotik committed:
more on lift and liftIO

fix link to point to latest on stackage
add note about equivalence between lift and liftIO
add example demonstrating the usefulness of liftIO

levinotik commented:
Thanks @snoyberg. I've added a note about that as well as an example demonstrating the need for and usefulness of `liftIO`

levinotik commented:
Also, please note I'm a relative noob with Haskell so I'm as much trying to learn as I am trying to help others :)

levinotik committed:
typo

levinotik committed:
formatting

levinotik committed:
change EitherT reference to ExceptT

snoyberg commented on content/monad-transformers.md:
```diff
@@ -23,13 +23,96 @@ A `Reader r a` in which the resulting `a` is wrapped in any other monad, i.e. `r
-A `State s a` in which the return value `a` and state `s` are wrapped in any other monad, i.e. `s -> m (a, s)`
+A `State s a` in which the return value and state, namely `(a, s)`, are wrapped in any other monad, i.e. `s -> m (a, s)`
 
-### EitherT
```
I actually would stick with the `EitherT`. I discussed it with others some time in the past, and there are still ways in which `EitherT` is superior.

levinotik replied:
Ah ok. I just thought it might be confusing given the focus on the transformers package later.

snoyberg replied:
That's a really good point. I agree with the change to `ExceptT`.

levinotik replied:
I think it'd be great to document those discussions though. I was surprised that what I assumed would be called `EitherT` in transformers was actually called `ExceptT` and even more confused that there was a separate `EitherT` package in addition!

snoyberg replied:
IMO (and the opinion of a few others), it's a really unfortunate outcome of some decisions in transformers 0.4. I agree that including a "should I use EitherT or ExceptT" is a good idea, but it's likely to be a very subjective section.

levinotik replied:
Understood. It seems the docs on "exceptions best practices" also included some subjective thoughts, but it's still very useful nonetheless! It's extremely helpful to at least be aware of the debate and well-reasoned arguments on both sides.

snoyberg replied:
I'm definitely in agreement here. I'm not saying "no subjective content" (since as you say I already wrote some!), but:
1. Objective content is better when possible
2. Subjective content should be called out as subjective/debatable

snoyberg commented on content/monad-transformers.md:
```diff
@@ -23,13 +23,96 @@ A `Reader r a` in which the resulting `a` is wrapped in any other monad, i.e. `r
+
+[transformers](https://www.stackage.org/package/transformers) is a widely used package which provides transformer versions of various monads. It also provides two useful classes, `MonadTrans` and `MonadIO`. 
+
+`MonadTrans` makes it easy to embed one monad into another. All of the transformers defined in the `transformers` package are instances `MonadTrans`.  
```
This phrasing is somewhat inaccurate, and possibly confusing. A very strict explanation is that "`MonadTrans` represents the ability to lift a monadic action from a base monad into a transformed monad," but that's a mouthful and pretty opaque for beginners. Perhaps "instances of `MonadTrans` are transformers which can be applied to other monads to create new monads," or something of that ilk?

snoyberg commented:
> please note I'm a relative noob with Haskell

I was unaware of that actually :) I gave a few more comments on some possible wording changes.

levinotik committed:
better explanation of lift and liftIO

change phrasing of lift explanation
add better motivation for liftIO

levinotik commented:
@snoyberg Thanks. I've changed the phrasing to use what you've suggested and significantly reworked the `liftIO` discussion, providing better motivation for it.

levinotik committed:
fix ghci code block formatting

levinotik committed:
typo

snoyberg commented:
Looks great, thanks! For the record, you should now have commit access to the repo, so feel free to add changes directly if you want. (I also don't mind reviewing PRs if you'd like the review.)
